File: generators/ee1a_generator.py
from datetime import datetime
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
from reportlab.lib.utils import ImageReader
from PIL import Image


class EE1AGenerator:
    def __init__(self, template_path="templates/EE-1a.pdf"):
        self.template_path = template_path

    def generate(self, client_record: dict, form_data: dict) -> tuple[str, BytesIO]:
        first_name = form_data.get("first_name", "")
        last_name = form_data.get("last_name", "")
        case_id = form_data.get("case_id", "")
        address_main = form_data.get("address_main", "")
        address_city = form_data.get("address_city", "")
        address_state = form_data.get("address_state", "")
        address_zip = form_data.get("address_zip", "")
        phone = form_data.get("phone", "")
        diagnoses = form_data.get("diagnoses", [])
        signature_file = form_data.get("signature_file")

        # Parse phone number
        phone_clean = phone.replace(".", "").replace("-", "").replace(" ", "")
        area_code = phone_clean[:3] if len(phone_clean) >= 3 else ""
        prefix = phone_clean[3:6] if len(phone_clean) >= 6 else ""
        line_number = phone_clean[6:10] if len(phone_clean) >= 10 else ""

        # Generate filename
        current_date = datetime.now().strftime("%m.%d.%y")
        if first_name and last_name:
            filename = f"EE1a_{first_name[0]}.{last_name}_{current_date}.pdf"
        else:
            filename = f"EE1a_Unknown_{current_date}.pdf"

        # Generate PDF
        pdf_bytes = self.draw_pdf(
            last_name,
            first_name,
            case_id,
            address_main,
            address_city,
            address_state,
            address_zip,
            area_code,
            prefix,
            line_number,
            diagnoses,
            signature_file,
        )
        return filename, pdf_bytes

    def draw_pdf(
        self,
        last_name: str,
        first_name: str,
        case_id: str,
        address_main: str,
        address_city: str,
        address_state: str,
        address_zip: str,
        area_code: str,
        prefix: str,
        line_number: str,
        diagnoses: list,
        signature_file,
    ) -> BytesIO:
        # Load the existing PDF
        base_pdf = PdfReader(self.template_path)
        base_page = base_pdf.pages[0]

        # Create overlay
        overlay_buffer = BytesIO()
        overlay = canvas.Canvas(overlay_buffer, pagesize=letter)
        overlay.setFont("Helvetica", 10)

        # Accurate coordinates based on EE-1a grid mapping

        # 1. Name (Last, First, Middle Initial) - separate fields like EE1
        overlay.drawString(25, 615, last_name)
        overlay.drawString(185, 615, first_name)

        # 2. Case ID Number
        overlay.drawString(400, 615, case_id)

        # 3. Address (Street, Apt. #, P.O. Box)
        overlay.drawString(25, 582, address_main)

        # 3. Address (City, State, ZIP Code) - separate fields like EE1
        overlay.drawString(25, 555, address_city)  # City
        overlay.drawString(215, 555, address_state)  # State
        overlay.drawString(255, 555, address_zip)  # ZIP

        # 4. Telephone Numbers
        # 4a. Home phone - position in the parentheses and dashes (using EE1 coordinates)
        if area_code:
            overlay.drawString(355, 583, area_code)  # Inside the ( ) for area code
        if prefix:
            overlay.drawString(390, 583, prefix)  # Before the dash
        if line_number:
            overlay.drawString(428, 583, line_number)  # After the dash

        # 5. Diagnoses - accurate positioning for fields a-e
        diagnosis_y_positions = [496, 479, 461, 443, 426]  # a, b, c, d, e
        for i, diagnosis in enumerate(diagnoses):
            if i < 5:  # Only 5 fields available (a-e)
                y_pos = diagnosis_y_positions[i]
                # Support both old format (diagnosis) and new format (diagnosis_text)
                diagnosis_text = diagnosis.get("diagnosis_text") or diagnosis.get("diagnosis", "")
                overlay.drawString(33, y_pos, diagnosis_text)

                # 6. Date of Diagnosis (Month, Day, Year)
                # Support both old format (date as datetime) and new format (diagnosis_date as string)
                date_value = diagnosis.get("diagnosis_date") or diagnosis.get("date")
                if date_value:
                    # If it's a string, parse it
                    if isinstance(date_value, str):
                        try:
                            date_obj = datetime.strptime(date_value, "%Y-%m-%d")
                        except ValueError:
                            continue
                    else:
                        date_obj = date_value

                    # Split into separate fields as shown in template
                    overlay.drawString(510, y_pos, f"{date_obj.month:02d}")  # Month
                    overlay.drawString(543, y_pos, f"{date_obj.day:02d}")  # Day
                    overlay.drawString(568, y_pos, f"{date_obj.year}")  # Year

        # Signature handling
        if signature_file:
            try:
                # Process the signature image
                signature_image = Image.open(signature_file)

                # Calculate scaling to fit signature area while maintaining aspect ratio
                # Signature area: approximately 200 width x 60 height points
                max_width, max_height = 200, 60

                # Get original dimensions
                orig_width, orig_height = signature_image.size

                # Calculate scale factor to fit within bounds while maintaining aspect ratio
                width_scale = max_width / orig_width
                height_scale = max_height / orig_height
                scale = min(width_scale, height_scale)

                # Only resize if image is larger than the target area
                if scale < 1:
                    new_width = int(orig_width * scale)
                    new_height = int(orig_height * scale)
                    signature_image = signature_image.resize(
                        (new_width, new_height),
                        Image.Resampling.LANCZOS
                    )

                # Convert to RGB if necessary (for PDF compatibility)
                if signature_image.mode in ("RGBA", "LA", "P"):
                    # Create white background for transparency
                    background = Image.new("RGB", signature_image.size, (255, 255, 255))
                    if signature_image.mode == "RGBA":
                        background.paste(signature_image, mask=signature_image.split()[3])
                    else:
                        background.paste(signature_image)
                    signature_image = background

                # Create temporary buffer for signature - use PNG for better quality
                sig_buffer = BytesIO()
                signature_image.save(sig_buffer, format="PNG", optimize=False)
                sig_buffer.seek(0)

                # Add signature to PDF - positioned in Claimant Signature field
                overlay.drawImage(
                    ImageReader(sig_buffer),
                    108,  # x position - aligned with other fields
                    165,  # y position - in signature area
                    width=signature_image.width,
                    height=signature_image.height,
                    preserveAspectRatio=True,
                )
            except Exception as e:
                # If signature processing fails, add text placeholder
                overlay.drawString(85, 185, "[Signature processing failed]")

        # Add current date next to signature (like EE1)
        current_date = datetime.now().strftime("%m/%d/%Y")
        overlay.drawString(385, 175, current_date)  # Date position next to signature

        overlay.save()
        overlay_buffer.seek(0)

        # Merge template onto overlay (so signature appears behind form elements)
        overlay_pdf = PdfReader(overlay_buffer)
        overlay_page = overlay_pdf.pages[0]

        output = PdfWriter()
        overlay_page.merge_page(base_page)
        output.add_page(overlay_page)

        result = BytesIO()
        output.write(result)
        result.seek(0)
        return result

File: web/src/lib/generators/ee1a-generator.ts
/**
 * EE-1A Form Generator
 * Converts Python ee1a_generator.py to TypeScript
 * NOTE: Includes image processing for signatures
 */

import { BaseGenerator } from './base-generator';
import { ClientRecord, GeneratorResult } from './types';
import { formatDateMMDDYY, formatDateMMDDYYYY, parsePhoneNumber } from './utils/formatters';
import { StandardFonts } from 'pdf-lib';
import sharp from 'sharp';

export interface EE1ADiagnosis {
  diagnosis_text?: string;
  diagnosis?: string; // Old format fallback
  diagnosis_date?: string; // ISO date string
  date?: string; // Old format fallback
}

export interface EE1AFormData {
  first_name: string;
  last_name: string;
  case_id: string;
  address_main: string;
  address_city: string;
  address_state: string;
  address_zip: string;
  phone: string;
  diagnoses: EE1ADiagnosis[];
  signature_file?: {
    data: string; // base64 encoded
  };
}

export class EE1AGenerator extends BaseGenerator {
  constructor() {
    super('EE-1a.pdf');
  }

  async generate(
    clientRecord: ClientRecord,
    doctor: string,
    formData: EE1AFormData
  ): Promise<GeneratorResult> {
    const firstName = formData.first_name || '';
    const lastName = formData.last_name || '';
    const caseId = formData.case_id || '';
    const addressMain = formData.address_main || '';
    const addressCity = formData.address_city || '';
    const addressState = formData.address_state || '';
    const addressZip = formData.address_zip || '';
    const phone = formData.phone || '';
    const diagnoses = formData.diagnoses || [];
    const signatureFile = formData.signature_file;

    // Parse phone number
    const { areaCode, prefix, lineNumber } = parsePhoneNumber(phone);

    // Generate filename
    const currentDate = formatDateMMDDYY();
    const filename =
      firstName && lastName
        ? `EE1a_${firstName[0]}.${lastName}_${currentDate}.pdf`
        : `EE1a_Unknown_${currentDate}.pdf`;

    // Load template
    const pdfDoc = await this.loadTemplate();
    const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
    const page = pdfDoc.getPage(0);

    page.setFont(font);
    page.setFontSize(10);

    // 1. Name (Last, First, Middle Initial)
    this.drawText(page, lastName, { x: 25, y: 615, size: 10 });
    this.drawText(page, firstName, { x: 185, y: 615, size: 10 });

    // 2. Case ID Number
    this.drawText(page, caseId, { x: 400, y: 615, size: 10 });

    // 3. Address (Street, Apt. #, P.O. Box)
    this.drawText(page, addressMain, { x: 25, y: 582, size: 10 });

    // 3. Address (City, State, ZIP Code)
    this.drawText(page, addressCity, { x: 25, y: 555, size: 10 });
    this.drawText(page, addressState, { x: 215, y: 555, size: 10 });
    this.drawText(page, addressZip, { x: 255, y: 555, size: 10 });

    // 4. Telephone Numbers - Home phone
    if (areaCode) this.drawText(page, areaCode, { x: 355, y: 583, size: 10 });
    if (prefix) this.drawText(page, prefix, { x: 390, y: 583, size: 10 });
    if (lineNumber) this.drawText(page, lineNumber, { x: 428, y: 583, size: 10 });

    // 5. Diagnoses
    const diagnosisYPositions = [496, 479, 461, 443, 426]; // a, b, c, d, e

    for (let i = 0; i < Math.min(diagnoses.length, 5); i++) {
      const diagnosis = diagnoses[i];
      const yPos = diagnosisYPositions[i];

      // Diagnosis text
      const diagnosisText =
        diagnosis.diagnosis_text || diagnosis.diagnosis || '';
      if (diagnosisText) {
        this.drawText(page, diagnosisText, { x: 33, y: yPos, size: 10 });
      }

      // 6. Date of Diagnosis (Month, Day, Year)
      const dateValue = diagnosis.diagnosis_date || diagnosis.date;
      if (dateValue) {
        try {
          const dateObj = new Date(dateValue);
          const month = String(dateObj.getMonth() + 1).padStart(2, '0');
          const day = String(dateObj.getDate()).padStart(2, '0');
          const year = String(dateObj.getFullYear());

          this.drawText(page, month, { x: 510, y: yPos, size: 10 });
          this.drawText(page, day, { x: 543, y: yPos, size: 10 });
          this.drawText(page, year, { x: 568, y: yPos, size: 10 });
        } catch (error) {
          console.error('[EE1A] Failed to parse diagnosis date:', error);
        }
      }
    }

    // Signature handling
    if (signatureFile?.data) {
      try {
        // Decode base64 image
        const imageBuffer = Buffer.from(signatureFile.data, 'base64');

        // Process with sharp: resize and convert to PNG
        const maxWidth = 200;
        const maxHeight = 60;

        const processedImage = await sharp(imageBuffer)
          .resize(maxWidth, maxHeight, {
            fit: 'inside',
            withoutEnlargement: true,
          })
          .flatten({ background: { r: 255, g: 255, b: 255 } }) // White background
          .png()
          .toBuffer();

        // Embed image in PDF
        const pngImage = await pdfDoc.embedPng(processedImage);
        const dims = pngImage.scale(1);

        page.drawImage(pngImage, {
          x: 108,
          y: 165,
          width: dims.width,
          height: dims.height,
        });
      } catch (error) {
        console.error('[EE1A] Signature processing failed:', error);
        this.drawText(page, '[Signature processing failed]', {
          x: 85,
          y: 185,
          size: 10,
        });
      }
    }

    // Add current date next to signature
    const formattedCurrentDate = formatDateMMDDYYYY();
    this.drawText(page, formattedCurrentDate, { x: 385, y: 175, size: 10 });

    // Save PDF
    const pdfBytes = await pdfDoc.save();

    return {
      filename,
      pdfBytes: Buffer.from(pdfBytes),
    };
  }
}
